### PR TITLE
Relationships Improvements

### PR DESCRIPTION
**Pull request by krasnoukhov**
- [x] Get rid of `URLForRelationship` in favor of relationship links
- [x] Fix `RelationshipRemoveOperation` to actually do the thing
- [x] Adjust payload format for all relationship operations according to the spec
- [x] Allow removing toOne relationship
- [x] Properly handle relationship persistence errors

### CODE REVIEW

**Comment by krasnoukhov** [comment]
Just added another commit that fixes a typo in `RelationshipRemoveOperation` and also makes it spec-compliant.

**Comment by wvteijlingen** [comment]
:+1: thanks, looking good! I think there are still a few cases that are not covered though:

1. Replacing a to-one relationship will null. `convertResourcesToLinkage` now always returns `[[String: String]]`. This forces the use of `.first` as you did, but also disallows returning null. We might need to change the system to allow three different linkage types: a dictionary, an array of dictionaries, or null.
2. `URLForRelationship` of router is not spec compliant anymore. It returns a `/links` URL whereas it should be a `/relationships` url. Fairly easy to change for now. Perhaps we should think about not using a hardcoded URL pattern here but instead use the `self` URL that is contained in the parent resource's link.

**Comment by krasnoukhov** [comment]
Ok, let me see if I can take care of those. Ran into `/links` issue as well, but went with creating custom router. Would be nice to get rid of that for sure.

**Comment by krasnoukhov** [comment]
Ok, I think the second point you've brought is addressed now. I've updated PR description for remaining issues which I'm going to handle in this scope. #54 is closed because it depends on the stuff I change here, and also I'm going to cover those cases with specs. #53 should be good to merge separately.

Please let me know what you think and if there are any additional concerns.

**Comment by wvteijlingen** [comment]
I'm a bit on the fence about adding a `linkURL` to a resource itself. Been thinking about it and I see three options now:

1. Add `linkURL` to `Resource`. Upside is that this is the easiest option to implement (as you just did). Downside is that we're adding some data to a resource that doesn't really belong to that resource. The same resource can be linked to from multiple parents, each through a different `linkURL`. But it's still the same resource. I foresee a problem here when a resource is included (side loaded), and referenced multiple times through different relationship URLs. We'd end up overwriting the URL on the linked resource.

2. Place an intermediate object between the parent and linked resource. Like we do now with `LinkedResourceCollection`. In this object we can store the `self` URL, as well as other link metadata. This is probably the cleanest way to represent a link in objects. Downside is that then every linked Resource must be expressed as a `LinkedResource`. That destroys type interference and makes for super ugly accessor syntax: `((post.author!.resource as! User).address!.resource as! Address)`, or something like that. Perhaps we could solve this via generics, but I believe KVC doesn't work with generics.

3. Store link data on the parent resource. Add a `relationships` field to `Resource` and store all raw relationship link in there. Upside is that we could store other things like this later as well. For example the original field values so we can do dirty checking.

None of these options are optimal, but my preference currently goes to the third option. I don't think the first one is possible due to overwriting URLs, and the second one defeats the Swift type system too much.

Let me know what you think.

**Comment by krasnoukhov** [comment]
Thanks for the details, I think third option looks reasonable for me. I can work on that later. Please let me know what you think of other issues I've outlined.

**Comment by krasnoukhov** [comment]
@wvteijlingen Ok, everything should be adressed. Please take a look. There's one controversial thing with deserializing save result for parent resource before relations are persisted, so implementation might be a bit ugly. Do you have any ideas on how to improve that?

File: Spine/Operation.swift [hunk 1]
```diff
@@ -292,11 +282,11 @@ class SaveOperation: ConcurrentOperation {
 					self.state = .Finished
 					return
 				}
-				
 			} else {
 				if let data = responseData where data.length > 0 {
 					do {
-						try self.serializer.deserializeData(data, mappingTargets: [self.resource])
+						try self.serializer.deserializeData(data, mappingTargets: nil)
```
Comment (wvteijlingen) [inline Spine/Operation.swift]: Perhaps I'm missing something but why is `mappingTargets` nil?

Comment (krasnoukhov) [inline Spine/Operation.swift]: Well, this one is just to test that deserialization throws no error. For resource, it's being applied in `deserializeIntoResource`. Any ideas on how to improve that?

**Comment by wvteijlingen** [comment]
I've looked through the changes and it's going in the right direction I think! However, there are a few points I think should be improved:

- The relationship links are currently retrieved in the operations themselves. I think this should still be done in the router. That way you can override the default behaviour if needed. The routers should ultimately be the definitive authority to build URLs.
- I think it's nicer to encapsulate the relationship links in a Relationship struct, containing the links and the linkage data. Such a struct could then also be reused in LinkedResourceCollection.
- `RelationshipToOneUpdateOperation` and `RelationshipToOneRemoveOperation` could be merged. They should be both issue a PATCH request, just using `null` when deleting as specified in http://jsonapi.org/format/#crud-updating-to-one-relationships.
- Not sure about this but perhaps even the to many and to one update operations could be merged. They both issue PATCH with new IDs.
- I don't get the use of `deserializeIntoResource`. Why not deserialise onto the resources immediately in the network callback?
- Al in all, I think the SaveOperation is a bit of a mess. Nothing to do with this PR, it was like that already. We should look into getting rid of all those result/state/return blocks. Probably also clean up the nested `statusCodeIsSuccess` + `if let data = responseData where data.length > 0` ifs.

**Comment by krasnoukhov** [comment]
Thanks for the reply.

> I think it's nicer to encapsulate the relationship links in a Relationship struct

This is a nice idea, but I'm not sure how to accomplish this properly. If I understand correctly, we have to mutate `fields` property of a specific resource in order to assign links dictionary to a `Relationship` instance. That's not possible given the current code that proxies `fields` to a `dynamicType`. Do you want me to change that?

> Not sure about this but perhaps even the to many and to one update operations could be merged

I think it would be better to keep these separate because of the different payload format. I just think it's clearer approach.

> I don't get the use of deserializeIntoResource. Why not deserialise onto the resources immediately in the network callback?

This is because of the order of operations. I noticed that while I actually tried to write relationship tests. The thing is that resource is saved and updated from a backend _before_ the relationship operations apply, which leads to replacing the local relationship changes with server-updated data, so relationship operations are skipped since there is basically no change. I'd like to improve the approach if you have any ideas. I was thinking maybe we could clone the resource, but that seems suboptimal. Another thing could be saving just the relationships data for future use, but that's also ugly I think.

Please let me know what you think so I can continue with changes.

**Comment by wvteijlingen** [comment]
> This is a nice idea, but I'm not sure how to accomplish this properly.

I mean creating as follows:

```swift
struct RelationshipData {
	var selfURL: NSURL?
	var relatedURL: NSURL?
	var data: [ResourceIdentifier]
}

public class Resource: NSObject, NSCoding {
	/// The relationships dictionary for this resource.
	var relationships: [String: RelationshipData]?
}

> I think it would be better to keep these separate because of the different payload format. I just think it's clearer approach.

Agreed for now, it's better to keep to-one and to-many separate.

> This is because of the order of operations.

Ah yes I see, never thought about that, good catch! Perhaps we can first update the relationships and then update the resource? Then we can just deserialise the data returned from the server since the relationships are already updated.

**Comment by wvteijlingen** [comment]
Could you create a new PR that merges into the `spine:relationships` branch? That way I can merge this while keeping it separate from master for now.

**Comment by krasnoukhov** [comment]
Ok sure. Sorry for the delays, I've got distracted by other projects. Opening a new PR now.

**Pull request closed without merging**